### PR TITLE
Recreate LDAP client object after authenticating

### PR DESCRIPTION
**Pull request by jberkenbilt**
In the flow of binding to ldap, searching for the user, binding again
as that user, and then finally binding back as the original bind user,
create a fresh LDAP client object before the final rebinding step.
Without this, the rebinding step fails when the ldap server is DUO's
ldap proxy.

I have been able to reproduce this behavior outside of vault in a simple, stand-alone program that uses go-ldap. I do not at this time know whether the problem is with go-ldap, with how vault is calling go-ldap, or with DUO's LDAP proxy, but other LDAP implementations that I have tried (in both C and Java) are able to talk with DUO's proxy without a problem.

Without this fix, the error `unable to read LDAP response packet: unexpected EOF` is returned by the call that does the rebinding.

I am going to open a separate issue with go-ldap and will link the two issues. This change, whether fix or workaround, allows vault's ldap backend to point to a DUO ldap proxy as its LDAP server. This is important for cases in which this is the only LDAP server available that vault can talk to. I hope you will consider its inclusion.

### CODE REVIEW

**Comment by jefferai** [comment]
I think the first thing to do would be to see if go-ldap authors have any response, as it sounds like something specific to that library and Duo's LDAP proxy. It would be nice to not force rebinding the client on every user of Vault if it's only Duo's LDAP proxy that is problematic (since we've never heard of this complaint before).

I would also encourage you to engage your Duo support contacts.

**Comment by jberkenbilt** [comment]
@jefferai I just now referenced this from the issue I opened with go-ldap. I will also reach out to DUO.

**Comment by jefferai** [comment]
This won't go into 1.0.0. If review goes well it might be in 1.0.1.

**Comment by jberkenbilt** [comment]
1.0.1 would still be great. Is there anything I have to do to get a review on this?

File: builtin/credential/ldap/backend.go [hunk 1]
```diff
@@ -118,9 +118,25 @@ func (b *backend) Login(ctx context.Context, req *logical.Request, username stri
 	if cfg.BindDN != "" && cfg.BindPassword != "" {
 		if err := c.Bind(cfg.BindDN, cfg.BindPassword); err != nil {
 			if b.Logger().IsDebug() {
-				b.Logger().Debug("error while attempting to re-bind with the BindDN User", "error", err)
+				b.Logger().Debug("error while attempting to re-bind with the BindDN User; trying with new connection", "error", err)
+			}
+			// Create a new client object to work around an issue with
```
Comment (vishalnayak) [inline builtin/credential/ldap/backend.go]: Can we link the issue in the comment here, which might potentially in future lead to removal of this block of code?

File: builtin/credential/ldap/backend.go [hunk 2]
```diff
@@ -118,9 +118,25 @@ func (b *backend) Login(ctx context.Context, req *logical.Request, username stri
...
+			// go-ldap and DUO's LDAP proxy. It is not known whether this is
+			// a bug with go-ldap or with DUO's proxy.
+			c, err = ldapClient.DialLDAP(cfg)
+			if err != nil {
+				return nil, logical.ErrorResponse(err.Error()), nil, nil
```
Comment (vishalnayak) [inline builtin/credential/ldap/backend.go]: This should be a `return nil, nil, nil, err`.

File: builtin/credential/ldap/backend.go [hunk 3]
```diff
@@ -118,9 +118,25 @@ func (b *backend) Login(ctx context.Context, req *logical.Request, username stri
...
+			}
+			if c == nil {
+				return nil, logical.ErrorResponse("invalid connection returned from LDAP dial"), nil, nil
```
Comment (vishalnayak) [inline builtin/credential/ldap/backend.go]: This should be a `return nil, nil, nil, errors.New("invalid connection returned from LDAP dial")`

**Comment by jberkenbilt** [comment]
@vishalnayak Thanks for the review. Comments addressed.

**Comment by jberkenbilt** [comment]
I pushed up a reworded comment that takes into consideration the new information from DUO. From my perspective, this is ready to go. Please let me know if there's anything else I need to do, if this is good now and will get merged at the appropriate time in your development cycle. Thanks!

**Comment by jberkenbilt** [comment]
It seems likely that we are going to be able to resolve this with a configuration change on the DUO side. In that case, should I close this, or is it still worth considering?

**Comment by jberkenbilt** [comment]
If all goes well, I should have an answer by tomorrow morning as to whether we can fix this with a config change.

**Comment by jberkenbilt** [comment]
It turns out that if you set `allow_unlimited_binds=true` in the configuration of the DUO LDAP proxy, then it works perfectly with the bind-search-bind-bind authentication path used by vault out of the box, making this patch unnecessary. There are a few other settings that have to be made for this to be secure, but the DUO docs are pretty clear. I mention this for the benefit of someone else who may find this.

Should we just close this, or is it still useful?

**Pull request closed without merging**